### DISCUSSION TITLE
"histogram" event is not being triggered when using an event listener in plugin

### DISCUSSION THREAD

**Discussion by servohatred** [artilleryio/artillery#2562] [Artillery]
Version info: 
```
2.0.7
```

Running this command: 
```
 artillery run mytestfile.yaml
```

I expected to see this happen:

`The histogram event to be fired and my callback to be executed`

Instead, this happened:

`the event was not fired
`

In my test processor(playwright engine) I've been using `events.emit("histogram", "customevent", duration)` to record my custom metrics, now I want to grab my test summary to fetch my event results programatically so I can send this data to a reporter. I am trying that by doing so(also got this from playwright reporter code in the publish metrics plugin):
` events.on('histogram', (name, value, metadata) => {console.log("this event was fired")} )`
this event never gets triggered even if it's displayed in the console reporter summary.
Also tried by using a similar approach I found in the publish metrics datadog plugin:
`events.on('stats', (stats) => {console.log(stats.reporter())})`
this event works and it's fired but all the reports I get from this approach are empty and I dont get any data related to my custom metrics(custom metrics, summaries and counters are always empty) . 

Is there a way to collect the data from  the test summary in the console output? I've been reading a report json file but due to concurrency issues with the files I need to find a way to send my data from an after hook or a plugin cleanup

**Comment by hassy** [comment]
hi @servohatred 👋 emitting metrics with `histogram` (and `counter`) is one-way - all emitted metrics are consumed and then aggregated by Artillery's metrics engine, and then aggregated reports are emitted via `stats` and `done` events to consumers (i.e. plugins).

Not sure why you're not receiving the `stats` event though - it's built in functionality that e.g. the `publish-metrics` plugin relies on, and should be working.

**Reply by hassy** [reply]
(moved to Discussions)

**Reply by servohatred** [reply]
Thanks for your answer. When I try listening to "done" event it works but sadly is not showing any custom event.  If I try with "stats" event i just get blank summaries
stats.reporter():
```
{
  customStats: {},
  counters: {},
  scenariosAvoided: 0,
  timestamp: 2024-03-12T18:42:50.000Z,
  scenariosCreated: 1,
  scenariosCompleted: 1,
  requestsCompleted: 0,
  latency: {},
  rps: { mean: 0, count: 0 },
  scenarioDuration: {
    min: 72809.8,
    max: 72809.8,
    count: 1,
    mean: 72809.8,
    p50: 72426.2,
    median: 72426.2,
    p75: 72426.2,
    p90: 72426.2,
    p95: 72426.2,
    p99: 72426.2,
    p999: 72426.2
  },
  scenarioCounts: { 'myscenario': 1 },
  errors: {},
  codes: {}
}
```

Should it be displayed on customstats? 
ps: my custom metric is displayed in the console reporter but not here

**Reply by bernardobridge** [reply]
Hey @servohatred 👋 . I believe you need to export `LEGACY_METRICS_FORMAT: false` from your Plugin, similar to [here](https://github.com/artilleryio/artillery/blob/ef2802b6cf174ff7969bbccf8f236ed54257e163/packages/artillery-plugin-publish-metrics/index.js#L24). Then the `stats` event should work.

We should change that behaviour to be false by default soon though!

**Reply by servohatred** [reply]
But I am using a custom plugin .... There is no LEGACY_METRICS_FORMAT. I literally created a plugin and put the listener in there

**Reply by bernardobridge** [reply]
Hi @servohatred , that's what I meant, you need to export that from your plugin, so when Artillery runs your plugin, it will know not to use the `LEGACY_METRICS_FORMAT`. For example:

```js
module.exports = {
    Plugin: ArtilleryHelloWorldPlugin,
    LEGACY_METRICS_FORMAT: false
};

function ArtilleryHelloWorldPlugin(script, events) {
  this.events = events;
  
  events.on('stats', (stats) => {
    console.log(stats)
  })

  return this;
}
```

We need to change that from defaulting to `true` on our side, it doesn't make sense anymore :)! But in the meantime, setting `LEGACY_METRICS_FORMAT` to `false` should work.

**Reply by bernardobridge** [reply]
Sorry @servohatred , the above snippet had a bug - it should be `console.log(stats)` not `console.log(stats.report())`. Let me know if it works!

**Reply by servohatred** [reply]
You are the best, thank you for all your answers, after using that flag it's showing all of them in both, "stats" and "done" events. Thank you so much
